lib/lambda_ethereum_consensus/p2p/block_downloader.ex

```diff
@@ -88,29 +96,43 @@ defmodule LambdaEthereumConsensus.P2P.BlockDownloader do
              @blocks_by_root_protocol_id,
              size_header <> compressed_payload
            ),
-         {:ok, payload} <- parse_chunk(response_chunk),
-         {:ok, block} <- decode_response(payload) do
-      {:ok, block}
+         {:ok, chunks} <- parse_response(response_chunk, length(roots)),
+         {:ok, blocks} <- decode_chunks(chunks) do
+      if length(blocks) != length(roots) do
+        {:error, "expected #{length(roots)} blocks, got #{length(blocks)}"}
+      else
+        {:ok, blocks}
+      end
     else
       {:error, reason} ->
         if retries > 0 do
-          Logger.debug("Retrying request for block with root #{Base.encode16(root)}")
-          request_block_by_root(root, retries - 1)
+          Logger.debug(
+            "Retrying request for block with roots #{Enum.map_join(roots, ", ", &Base.encode16/1)}"
+          )
+
+          request_blocks_by_root(roots, retries - 1)
         else
           {:error, reason}
         end
     end
   end
 
-  defp parse_chunk(response_chunk) do
+  @spec parse_response(binary, integer) ::
+          {:ok, [binary()]} | {:error, binary()}
+  def parse_response(response_chunk, count) do
     fork_context = @fork_context
 
     case response_chunk do
-      "" ->
+      <<>> ->
         {:error, "unexpected EOF"}
 
-      <<0, ^fork_context::binary-size(4)>> <> chunk ->
-        {:ok, chunk}
+      <<0, ^fork_context::binary-size(4)>> <> rest ->
+        result = rest |> :binary.split(<<0, fork_context::binary-size(4)>>)
```

Comment:
You need to specify the `:global` option for it to work on all matches.
```suggestion
        result = rest |> :binary.split(<<0, fork_context::binary-size(4)>>, [:global])
```